### PR TITLE
feat(quic): add support for listening on many addresses

### PR DESCRIPTION
**Pull request by vladopajic**
## Summary
adding ability for quic transport to listen to many addresses simultaneously
for commitment: `ift-ts:p2p:ift:2026q2-nimlibp2p-maintenance:quic-maddrs`

## Affected Areas

- [x] Transports (quic)

## Compatibility & Downstream Validation
no impact 

## Impact on Library Users
no impact 

## Risk Assessment
no impact 

## References
closes: https://github.com/vacp2p/nim-libp2p/issues/1663

### CODE REVIEW

File: libp2p/transports/quictransport.nim [hunk 1]
```diff
@@ -328,17 +334,25 @@ method start*(
...
-  if not transport.listener.isNil:
+  for listener in transport.listeners:
     try:
-      await transport.listener.stop()
+      await listener.stop()
     except CatchableError as exc:
       trace "Error shutting down Quic transport", description = exc.msg
-    transport.listener = nil
+  transport.listeners = @[]
+  transport.acceptFuts = @[]
```
Comment (richard-ramos) [inline libp2p/transports/quictransport.nim:355]: What happens if you call stop() and there's a future waiting in line 414?
Wouldnt the code fail in line 426:427 since the index returned by the find be -1? and that not being a valid index for self.acceptedFuts?

IMO in stop we should probably do something like this in the stop() function:
https://github.com/vacp2p/nim-libp2p/blob/master/libp2p/transports/tcptransport.nim#L178-L182

Comment (vladopajic) [reply libp2p/transports/quictransport.nim:355]: > Wouldnt the code fail in line 426:427 since the index returned by the find be -1?

this code will never return `-1`. it will always return some future that could be found in array. or, else, awaiting is cancelled. there are no other outcomes possible. 

`await one(..)` returns first "completed" future regardless of how it completed. in case when it is stopped this future will be finished with an exception exception. so we will be able to get future from acceptedFuts array. and if we then await on this future it will throw "quic stopped error".

[the remark this answers is not in the document]
Comment (vladopajic) [reply libp2p/transports/quictransport.nim:355]: no, `Listener.accept()` will be stopped if listener is stopped.

Comment (vladopajic) [reply libp2p/transports/quictransport.nim:355]: anyway i have added issue https://github.com/vacp2p/nim-libp2p/issues/2194 that will test these edge cases

we should test this across all transports

Comment (vladopajic) [reply libp2p/transports/quictransport.nim:355]: aha only issue might be this: `stop() clears acceptFuts` https://github.com/vacp2p/nim-libp2p/pull/2193#discussion_r2995375786

File: libp2p/transports/quictransport.nim [hunk 2]
```diff
@@ -389,11 +403,32 @@ method accept*(
...
+      for fut in self.acceptFuts:
+        if not fut.finished():
+          fut.cancelSoon()
+      raise exc
+
+  # becasue some listener has accepted we need to run 
+  # accept manually in the place for this listner again 
+  # so that it keeps accepting for future method calls
+  let index = self.acceptFuts.find(finished)
+  self.acceptFuts[index] = self.listeners[index].accept()
```
[the remark this answers is not in the document]
Comment (richard-ramos) [reply libp2p/transports/quictransport.nim:431]: How come? -1 is a valid return value for find when a value does not exist in a sequence:
https://nim-lang.org/docs/system.html#find%2CT%2CS

[the remark this answers is not in the document]
Comment (vladopajic) [reply libp2p/transports/quictransport.nim:431]: this is correct observation

[the remark this answers is not in the document]
Comment (richard-ramos) [reply libp2p/transports/quictransport.nim:431]: https://play.nim-lang.org/#pasty=ZNzIWRxi
proof of -1
